// next/dist/lib/router/router.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _slicedToArray2 = require('babel-runtime/helpers/slicedToArray');

var _slicedToArray3 = _interopRequireDefault(_slicedToArray2);

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _set = require('babel-runtime/core-js/set');

var _set2 = _interopRequireDefault(_set);

var _defineProperty2 = require('babel-runtime/helpers/defineProperty');

var _defineProperty3 = _interopRequireDefault(_defineProperty2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _url = require('url');

var _events = require('events');

var _evalScript = require('../eval-script');

var _evalScript2 = _interopRequireDefault(_evalScript);

var _shallowEquals = require('../shallow-equals');

var _shallowEquals2 = _interopRequireDefault(_shallowEquals);

var _pQueue = require('../p-queue');

var _pQueue2 = _interopRequireDefault(_pQueue);

var _utils = require('../utils');

var _ = require('./');

var _unfetch = require('unfetch');

var _unfetch2 = _interopRequireDefault(_unfetch);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

if (typeof window !== 'undefined' && typeof navigator.serviceWorker !== 'undefined') {
  navigator.serviceWorker.getRegistrations().then(function (registrations) {
    registrations.forEach(function (registration) {
      if (!registration.active) {
        return;
      }

      if (registration.active.scriptURL.indexOf('_next-prefetcher.js') === -1) {
        return;
      }

      console.warn('Unregistered deprecated \'next/prefetch\' ServiceWorker. See https://github.com/zeit/next.js#prefetching-pages');
      registration.unregister();
    });
  });
}

var Router = function (_EventEmitter) {
  (0, _inherits3.default)(Router, _EventEmitter);

  function Router(pathname, query, as) {
    var _ref = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : {},
        Component = _ref.Component,
        ErrorComponent = _ref.ErrorComponent,
        err = _ref.err;

    (0, _classCallCheck3.default)(this, Router);

    // represents the current component key
    var _this = (0, _possibleConstructorReturn3.default)(this, (Router.__proto__ || (0, _getPrototypeOf2.default)(Router)).call(this));

    _this.route = toRoute(pathname);

    // set up the component cache (by route keys)
    _this.components = (0, _defineProperty3.default)({}, _this.route, { Component: Component, err: err });

    // contain a map of promise of fetch routes
    _this.fetchingRoutes = {};

    _this.prefetchQueue = new _pQueue2.default({ concurrency: 2 });
    _this.ErrorComponent = ErrorComponent;
    _this.pathname = pathname;
    _this.query = query;
    _this.as = as;
    _this.subscriptions = new _set2.default();
    _this.componentLoadCancel = null;
    _this.onPopState = _this.onPopState.bind(_this);

    if (typeof window !== 'undefined') {
      // in order for `e.state` to work on the `onpopstate` event
      // we have to register the initial route upon initialization
      _this.changeState('replaceState', (0, _url.format)({ pathname: pathname, query: query }), (0, _utils.getURL)());

      window.addEventListener('popstate', _this.onPopState);
    }
    return _this;
  }

  (0, _createClass3.default)(Router, [{
    key: 'onPopState',
    value: function () {
      var _ref2 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee(e) {
        var pathname, query, _e$state, url, as;

        return _regenerator2.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                if (e.state) {
                  _context.next = 4;
                  break;
                }

                // We get state as undefined for two reasons.
                //  1. With older safari (< 8) and older chrome (< 34)
                //  2. When the URL changed with #
                //
                // In the both cases, we don't need to proceed and change the route.
                // (as it's already changed)
                // But we can simply replace the state with the new changes.
                // Actually, for (1) we don't need to nothing. But it's hard to detect that event.
                // So, doing the following for (1) does no harm.
                pathname = this.pathname, query = this.query;

                this.changeState('replaceState', (0, _url.format)({ pathname: pathname, query: query }), (0, _utils.getURL)());
                return _context.abrupt('return');

              case 4:
                _e$state = e.state, url = _e$state.url, as = _e$state.as;

                this.replace(url, as);

              case 6:
              case 'end':
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function onPopState(_x2) {
        return _ref2.apply(this, arguments);
      }

      return onPopState;
    }()
  }, {
    key: 'update',
    value: function update(route, Component) {
      var data = this.components[route] || {};
      var newData = (0, _extends3.default)({}, data, { Component: Component });
      this.components[route] = newData;

      if (route === this.route) {
        this.notify(newData);
      }
    }
  }, {
    key: 'reload',
    value: function () {
      var _ref3 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee2(route) {
        var url, _parse, pathname, query, _ref4, data, props, error;

        return _regenerator2.default.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                delete this.components[route];
                delete this.fetchingRoutes[route];

                if (!(route !== this.route)) {
                  _context2.next = 4;
                  break;
                }

                return _context2.abrupt('return');

              case 4:
                url = window.location.href;
                _parse = (0, _url.parse)(url, true), pathname = _parse.pathname, query = _parse.query;


                this.emit('routeChangeStart', url);
                _context2.next = 9;
                return this.getRouteInfo(route, pathname, query, url);

              case 9:
                _ref4 = _context2.sent;
                data = _ref4.data;
                props = _ref4.props;
                error = _ref4.error;

                if (!(error && error.cancelled)) {
                  _context2.next = 15;
                  break;
                }

                return _context2.abrupt('return');

              case 15:

                this.notify((0, _extends3.default)({}, data, { props: props }));

                if (!error) {
                  _context2.next = 19;
                  break;
                }

                this.emit('routeChangeError', error, url);
                throw error;

              case 19:

                this.emit('routeChangeComplete', url);

              case 20:
              case 'end':
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function reload(_x3) {
        return _ref3.apply(this, arguments);
      }

      return reload;
    }()
  }, {
    key: 'back',
    value: function back() {
      window.history.back();
    }
  }, {
    key: 'push',
    value: function push(url) {
      var as = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : url;

      return this.change('pushState', url, as);
    }
  }, {
    key: 'replace',
    value: function replace(url) {
      var as = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : url;

      return this.change('replaceState', url, as);
    }
  }, {
    key: 'change',
    value: function () {
      var _ref5 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee3(method, url, as) {
        var _parse2, pathname, query, route, _ref6, data, props, error, hash;

        return _regenerator2.default.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                this.abortComponentLoad(as);
                _parse2 = (0, _url.parse)(url, true), pathname = _parse2.pathname, query = _parse2.query;

                // If the url change is only related to a hash change
                // We should not proceed. We should only replace the state.

                if (!this.onlyAHashChange(as)) {
                  _context3.next = 5;
                  break;
                }

                this.changeState('replaceState', url, as);
                return _context3.abrupt('return');

              case 5:

                // If asked to change the current URL we should reload the current page
                // (not location.reload() but reload getInitalProps and other Next.js stuffs)
                // We also need to set the method = replaceState always
                // as this should not go into the history (That's how browsers work)
                if (!this.urlIsNew(pathname, query)) {
                  method = 'replaceState';
                }

                route = toRoute(pathname);


                this.emit('routeChangeStart', as);
                _context3.next = 10;
                return this.getRouteInfo(route, pathname, query, as);

              case 10:
                _ref6 = _context3.sent;
                data = _ref6.data;
                props = _ref6.props;
                error = _ref6.error;

                if (!(error && error.cancelled)) {
                  _context3.next = 16;
                  break;
                }

                return _context3.abrupt('return', false);

              case 16:

                this.changeState(method, url, as);
                hash = window.location.hash.substring(1);


                this.route = route;
                this.set(pathname, query, as, (0, _extends3.default)({}, data, { props: props, hash: hash }));

                if (!error) {
                  _context3.next = 23;
                  break;
                }

                this.emit('routeChangeError', error, as);
                throw error;

              case 23:

                this.emit('routeChangeComplete', as);
                return _context3.abrupt('return', true);

              case 25:
              case 'end':
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function change(_x6, _x7, _x8) {
        return _ref5.apply(this, arguments);
      }

      return change;
    }()
  }, {
    key: 'changeState',
    value: function changeState(method, url, as) {
      if (method !== 'pushState' || (0, _utils.getURL)() !== as) {
        window.history[method]({ url: url, as: as }, null, as);
      }
    }
  }, {
    key: 'getRouteInfo',
    value: function () {
      var _ref7 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee4(route, pathname, query, as) {
        var routeInfo, _routeInfo$data, Component, err, jsonPageRes, ctx, _Component, _ctx;

        return _regenerator2.default.wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                routeInfo = {};
                _context4.prev = 1;
                _context4.next = 4;
                return this.fetchComponent(route, as);

              case 4:
                routeInfo.data = _context4.sent;

                if (routeInfo.data) {
                  _context4.next = 7;
                  break;
                }

                return _context4.abrupt('return', null);

              case 7:
                _routeInfo$data = routeInfo.data, Component = _routeInfo$data.Component, err = _routeInfo$data.err, jsonPageRes = _routeInfo$data.jsonPageRes;
                ctx = { err: err, pathname: pathname, query: query, jsonPageRes: jsonPageRes };
                _context4.next = 11;
                return this.getInitialProps(Component, ctx);

              case 11:
                routeInfo.props = _context4.sent;
                _context4.next = 26;
                break;

              case 14:
                _context4.prev = 14;
                _context4.t0 = _context4['catch'](1);

                if (!_context4.t0.cancelled) {
                  _context4.next = 18;
                  break;
                }

                return _context4.abrupt('return', { error: _context4.t0 });

              case 18:
                _Component = this.ErrorComponent;

                routeInfo.data = { Component: _Component, err: _context4.t0 };
                _ctx = { err: _context4.t0, pathname: pathname, query: query };
                _context4.next = 23;
                return this.getInitialProps(_Component, _ctx);

              case 23:
                routeInfo.props = _context4.sent;


                routeInfo.error = _context4.t0;
                console.error(_context4.t0);

              case 26:
                return _context4.abrupt('return', routeInfo);

              case 27:
              case 'end':
                return _context4.stop();
            }
          }
        }, _callee4, this, [[1, 14]]);
      }));

      function getRouteInfo(_x9, _x10, _x11, _x12) {
        return _ref7.apply(this, arguments);
      }

      return getRouteInfo;
    }()
  }, {
    key: 'set',
    value: function set(pathname, query, as, data) {
      this.pathname = pathname;
      this.query = query;
      this.as = as;
      this.notify(data);
    }
  }, {
    key: 'onlyAHashChange',
    value: function onlyAHashChange(as) {
      if (!this.as) return false;

      var _as$split = this.as.split('#'),
          _as$split2 = (0, _slicedToArray3.default)(_as$split, 1),
          oldUrlNoHash = _as$split2[0];

      var _as$split3 = as.split('#'),
          _as$split4 = (0, _slicedToArray3.default)(_as$split3, 2),
          newUrlNoHash = _as$split4[0],
          newHash = _as$split4[1];

      // If the urls are change, there's more than a hash change


      if (oldUrlNoHash !== newUrlNoHash) {
        return false;
      }

      // If there's no hash in the new url, we can't consider it as a hash change
      if (!newHash) {
        return false;
      }

      // Now there's a hash in the new URL.
      // We don't need to worry about the old hash.
      return true;
    }
  }, {
    key: 'urlIsNew',
    value: function urlIsNew(pathname, query) {
      return this.pathname !== pathname || !(0, _shallowEquals2.default)(query, this.query);
    }
  }, {
    key: 'prefetch',
    value: function () {
      var _ref8 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee5(url) {
        var _this2 = this;

        var _parse3, pathname, route;

        return _regenerator2.default.wrap(function _callee5$(_context5) {
          while (1) {
            switch (_context5.prev = _context5.next) {
              case 0:
                if (!(process.env.NODE_ENV === 'development')) {
                  _context5.next = 2;
                  break;
                }

                return _context5.abrupt('return');

              case 2:
                _parse3 = (0, _url.parse)(url), pathname = _parse3.pathname;
                route = toRoute(pathname);
                return _context5.abrupt('return', this.prefetchQueue.add(function () {
                  return _this2.fetchRoute(route);
                }));

              case 5:
              case 'end':
                return _context5.stop();
            }
          }
        }, _callee5, this);
      }));

      function prefetch(_x13) {
        return _ref8.apply(this, arguments);
      }

      return prefetch;
    }()
  }, {
    key: 'fetchComponent',
    value: function () {
      var _ref9 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee6(route, as) {
        var data, cancelled, cancel, jsonPageRes, jsonData, error, newData, _error;

        return _regenerator2.default.wrap(function _callee6$(_context6) {
          while (1) {
            switch (_context6.prev = _context6.next) {
              case 0:
                data = this.components[route];

                if (!data) {
                  _context6.next = 3;
                  break;
                }

                return _context6.abrupt('return', data);

              case 3:
                cancelled = false;

                cancel = this.componentLoadCancel = function () {
                  cancelled = true;
                };

                _context6.next = 7;
                return this.fetchRoute(route);

              case 7:
                jsonPageRes = _context6.sent;
                _context6.next = 10;
                return jsonPageRes.json();

              case 10:
                jsonData = _context6.sent;

                if (!jsonData.buildIdMismatch) {
                  _context6.next = 16;
                  break;
                }

                (0, _._notifyBuildIdMismatch)(as);

                error = Error('Abort due to BUILD_ID mismatch');

                error.cancelled = true;
                throw error;

              case 16:
                newData = (0, _extends3.default)({}, loadComponent(jsonData), {
                  jsonPageRes: jsonPageRes
                });

                if (!cancelled) {
                  _context6.next = 21;
                  break;
                }

                _error = new Error('Abort fetching component for route: "' + route + '"');

                _error.cancelled = true;
                throw _error;

              case 21:

                if (cancel === this.componentLoadCancel) {
                  this.componentLoadCancel = null;
                }

                this.components[route] = newData;
                return _context6.abrupt('return', newData);

              case 24:
              case 'end':
                return _context6.stop();
            }
          }
        }, _callee6, this);
      }));

      function fetchComponent(_x14, _x15) {
        return _ref9.apply(this, arguments);
      }

      return fetchComponent;
    }()
  }, {
    key: 'getInitialProps',
    value: function () {
      var _ref10 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee7(Component, ctx) {
        var cancelled, cancel, props, err;
        return _regenerator2.default.wrap(function _callee7$(_context7) {
          while (1) {
            switch (_context7.prev = _context7.next) {
              case 0:
                cancelled = false;

                cancel = function cancel() {
                  cancelled = true;
                };

                this.componentLoadCancel = cancel;

                _context7.next = 5;
                return (0, _utils.loadGetInitialProps)(Component, ctx);

              case 5:
                props = _context7.sent;


                if (cancel === this.componentLoadCancel) {
                  this.componentLoadCancel = null;
                }

                if (!cancelled) {
                  _context7.next = 11;
                  break;
                }

                err = new Error('Loading initial props cancelled');

                err.cancelled = true;
                throw err;

              case 11:
                return _context7.abrupt('return', props);

              case 12:
              case 'end':
                return _context7.stop();
            }
          }
        }, _callee7, this);
      }));

      function getInitialProps(_x16, _x17) {
        return _ref10.apply(this, arguments);
      }

      return getInitialProps;
    }()
  }, {
    key: 'fetchRoute',
    value: function () {
      var _ref11 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee8(route) {
        var promise, res;
        return _regenerator2.default.wrap(function _callee8$(_context8) {
          while (1) {
            switch (_context8.prev = _context8.next) {
              case 0:
                promise = this.fetchingRoutes[route];

                if (!promise) {
                  promise = this.fetchingRoutes[route] = this.doFetchRoute(route);
                }

                _context8.next = 4;
                return promise;

              case 4:
                res = _context8.sent;
                return _context8.abrupt('return', res.clone());

              case 6:
              case 'end':
                return _context8.stop();
            }
          }
        }, _callee8, this);
      }));

      function fetchRoute(_x18) {
        return _ref11.apply(this, arguments);
      }

      return fetchRoute;
    }()
  }, {
    key: 'doFetchRoute',
    value: function doFetchRoute(route) {
      var buildId = window.__NEXT_DATA__.buildId;

      var url = '/_next/' + encodeURIComponent(buildId) + '/pages' + route;
      return (0, _unfetch2.default)(url, {
        method: 'GET',
        headers: { 'Accept': 'application/json' }
      });
    }
  }, {
    key: 'abortComponentLoad',
    value: function abortComponentLoad(as) {
      if (this.componentLoadCancel) {
        this.emit('routeChangeError', new Error('Route Cancelled'), as);
        this.componentLoadCancel();
        this.componentLoadCancel = null;
      }
    }
  }, {
    key: 'notify',
    value: function notify(data) {
      this.subscriptions.forEach(function (fn) {
        return fn(data);
      });
    }
  }, {
    key: 'subscribe',
    value: function subscribe(fn) {
      var _this3 = this;

      this.subscriptions.add(fn);
      return function () {
        return _this3.subscriptions.delete(fn);
      };
    }
  }]);
  return Router;
}(_events.EventEmitter);

exports.default = Router;


function toRoute(path) {
  return path.replace(/\/$/, '') || '/';
}

function loadComponent(jsonData) {
  var module = (0, _evalScript2.default)(jsonData.component);
  var Component = module.default || module;

  return { Component: Component, err: jsonData.err };
}
